lib/floki.ex

```diff
@@ -153,6 +153,59 @@ defmodule Floki do
     Enum.map(results, fn(html_node) -> HTMLTree.to_tuple(tree, html_node) end)
   end
 
+  defp add_nodes_to_tree(tree, [html_node]) do
```

Comment:
Can you move all the auxiliar functions (private ones) to the bottom of the public `attr/4`?
It's easier to understand what the role of the functions after read the public function that uses them.